### PR TITLE
#418: BaseService improvement

### PR DESCRIPTION
**Pull request by heshamMassoud**
#### Summary
Addresses #418

- Tests
    - [x] Unit 
    - [x] Integration
- [x] Documentation
- [x] Add Release Notes entry.

### CODE REVIEW

**Comment by heshamMassoud** [comment]
Hi @jarzynp! Thanks again for your contribution. I have reviewed the PR and added some refactoring. Feel free to criticise if you have objections on any change :)

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 1]
```diff
@@ -1,32 +1,50 @@
...
 import static org.apache.commons.lang3.StringUtils.isBlank;
 
 /**
  * @param <T> Resource Draft (e.g. {@link io.sphere.sdk.products.ProductDraft},
- *  {@link io.sphere.sdk.categories.CategoryDraft}, etc..
+ *            {@link io.sphere.sdk.categories.CategoryDraft}, etc..
  * @param <U> Resource (e.g. {@link io.sphere.sdk.products.Product}, {@link io.sphere.sdk.categories.Category}, etc..
  * @param <S> Subclass of {@link BaseSyncOptions}
+ * @param <Q> Query (e.g. {@link io.sphere.sdk.products.queries.ProductQuery},
+ *            {@link io.sphere.sdk.categories.queries.CategoryQuery}, etc..
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]: why not closing the brackets? 😄 
```suggestion
 *            {@link io.sphere.sdk.categories.queries.CategoryQuery}, etc.. )
```

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 2]
```diff
@@ -135,4 +153,138 @@
...
+    /**
+     * Given a {@code key}, if it is blank (null/empty), a completed future with an empty optional is returned.
+     * This method then checks if the cached map of resource keys -&gt; ids contains the key. If it does, then an
+     * optional containing the mapped id is returned. If the cache doesn't contain the key; this method attempts to
+     * fetch the id of the key from the CTP project, caches it and returns a
+     * {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt; in which the result of it's completion
+     * could contain an {@link Optional} with the id inside of it or an empty {@link Optional} if no resource
+     * was found in the CTP project with this key.
+     *
+     * @param key the key by which a resource id should be fetched from the CTP project.
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]:
```suggestion
     * @param key           the key by which a resource id should be fetched from the CTP project.
```

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 3]
```diff
@@ -135,4 +153,138 @@
...
+        return CtpQueryUtils
+            .queryAll(syncOptions.getCtpClient(), querySupplier.get(), pageConsumer)
+            .thenApply(result -> Optional.ofNullable(keyToIdCache.get(key)));
+    }
+
+    /**
+     * Given a set of keys this method caches a mapping of the keys to ids of such keys only for the keys which are
+     * not already in the cache.
+     *
+     * @param keys keys to cache.
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]:
```suggestion
     * @param keys            keys to cache.
```

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 4]
```diff
@@ -135,4 +153,138 @@
...
+            .queryAll(syncOptions.getCtpClient(), keysQueryMapper.apply(keysNotCached), pageConsumer)
+            .thenApply(result -> keyToIdCache);
+    }
+
+    /**
+     * Given a {@link Set} of resource keys, this method fetches a set of all the resources, matching this given set of
+     * keys in the CTP project, defined in an injected {@link SphereClient}. A mapping of the key to the id
+     * of the fetched resources is persisted in an in-memory map.
+     *
+     * @param keys  set of state keys to fetch matching states by
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]:
```suggestion
     * @param          keys set of state keys to fetch matching states by
```

File: docs/RELEASE_NOTES.md [hunk 5]
```diff
@@ -33,9 +33,20 @@
...
 
+- ✨ **Enhancements** (2)
+    - **Product Sync** - Only cache the needed keys of `Category`, `ProductType` and `Type` references instead of 
+    caching all keys of such resources. [#418](https://github.com/commercetools/commercetools-sync-java/issues/418) 
+    
+    **Note**: This might have performance implications on the product sync, since now every non cached key-id entry will be individually fetched. 
+    However, issue [#235](https://github.com/commercetools/commercetools-sync-java/issues/235) should address this.
+    
+    - **Commons** - Refactor duplicate implementations in concrete services and generalise it in the `BaseService`. 
+    Thanks @jarzynp for the contributions! [#418](https://github.com/commercetools/commercetools-sync-java/issues/418)
```
Comment (ahmetoz) [inline docs/RELEASE_NOTES.md]:
```suggestion
    Thanks, @jarzynp for the contributions! [#418](https://github.com/commercetools/commercetools-sync-java/issues/418)
```

File: src/test/java/com/commercetools/sync/services/impl/BaseServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+        final Optional<Product> optional = service.fetchProduct(key).toCompletableFuture().join();
+
+        //assertions
+        assertThat(optional).isEmpty();
+        verify(client, never()).execute(any());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    void fetchResource_WithKey_ShouldFetchResource() {
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/services/impl/BaseServiceImplTest.java]:
```suggestion
    void fetchResource_WithExistingProductKey_ShouldFetchResource() {
```

File: src/test/java/com/commercetools/sync/services/impl/BaseServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+        //test
+        final Optional<String> result = service.getIdFromCacheOrFetch(key).toCompletableFuture().join();
+
+        //assertions
+        assertThat(result).contains(id);
+    }
+
+    @Test
+    void fetchCachedResourceId_WithCachedResource_ShouldReturnResourceIdWithoutMakingRequest() {
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/services/impl/BaseServiceImplTest.java]:
```suggestion
    void fetchCachedResourceId_WithCachedKey_ShouldReturnResourceIdWithoutMakingRequest() {
```

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 1 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: @heshamMassoud  Please update others too

**Comment by heshamMassoud** [comment]
@jarzynp This is now released in [`1.7.0`](https://github.com/commercetools/commercetools-sync-java/blob/master/docs/RELEASE_NOTES.md#170----jan-7-2020) Thanks for contribution.

**Pull request merged**